packages/theme/src/cli/utilities/theme-environment/theme-environment.ts

```diff
@@ -28,46 +28,73 @@ export function setupDevServer(theme: Theme, ctx: DevServerContext) {
 function ensureThemeEnvironmentSetup(theme: Theme, ctx: DevServerContext) {
   const remoteChecksumsPromise = fetchChecksums(theme.id, ctx.session)
 
-  const reconcilePromise = remoteChecksumsPromise.then((remoteChecksums) =>
-    ctx.options.themeEditorSync
-      ? reconcileAndPollThemeEditorChanges(theme, ctx.session, remoteChecksums, ctx.localThemeFileSystem, {
-          noDelete: ctx.options.noDelete,
+  const editorSyncPromise = remoteChecksumsPromise.then((remoteChecksums) =>
+    handleThemeEditorSync(theme, ctx, remoteChecksums),
+  )
+
+  const themeSetupPromise = editorSyncPromise.then(
+    ({updatedRemoteChecksums, reconciliationFinishedPromise, readyForReconciliationPromise}) => {
+      const uploadPromise = reconciliationFinishedPromise.then(() =>
```

Comment:
Great question! We wait for the download to finish to prevent a race condition between this scenario
-> Download changes on `sections/example.json` from remote theme
-> At the same time, we may upload our local version of `sections/example.json` if the fileSystem hasn't been updated yet